qmanager/policies/base/queue_policy_base.hpp

```diff
@@ -526,14 +530,34 @@ class queue_policy_base_t : public resource_model::queue_adapter_base_t {
             case job_state_kind_t::ALLOC_RUNNING:
                 // deliberately fall through
             case job_state_kind_t::RUNNING:
-                if ((rc = cancel (h, job_it->second->id, R, true, full_removal) != 0))
-                    break;
-                if (full_removal) {
+                if (cancel (h, job_it->second->id, R, true, full_removal) != 0) {
+                    flux_log_error (flux_h,
+                                    "%s: .free RPC partial cancel failed for jobid "
+                                    "%jd",
+                                    __FUNCTION__,
+                                    static_cast<intmax_t> (id));
+                    errno = EPROTO;
+                    goto out;
+                }
+                if (full_removal || final) {
                     m_alloced.erase (job_it->second->t_stamps.running_ts);
                     m_running.erase (job_it->second->t_stamps.running_ts);
                     job_it->second->t_stamps.complete_ts = m_cq_cnt++;
                     job_it->second->state = job_state_kind_t::COMPLETE;
                     m_jobs.erase (job_it);
+                    if (final && !full_removal) {
+                        // This condition indicates a discrepancy between core and sched.
+                        // Run a full cancel to clean up all remaining allocated resources
+                        cancel (h, job_it->second->id, true);
+                        flux_log_error (flux_h,
+                                        "%s: Final .free RPC failed to remove all resources for "
+                                        "jobid "
+                                        "%jd",
+                                        __FUNCTION__,
+                                        static_cast<intmax_t> (id));
+                        errno = EPROTO;
+                        goto out;
```

Comment:
I may be misreading the logic, but it seems like if the second cancel succeeds,
it should be logged but probably shouldn't return error? Feel free to push back
on this one since I'm not sure what effects are at the callsite outside this but
if the resources are freed that seems substantially less broken than if some of
them couldn't be removed.

```suggestion
                        flux_log_error (flux_h,
                                        "%s: Final .free RPC failed to remove all resources for "
                                        "jobid "
                                        "%jd",
                                        __FUNCTION__,
                                        static_cast<intmax_t> (id));

                        if (cancel (h, job_it->second->id, true) != 0) {
                            errno = EPROTO;
                            goto out;
                        }
```

Reply:
> I may be misreading the logic, but it seems like if the second cancel succeeds,
it should be logged but probably shouldn't return error? 

I didn't write this in a very clear fashion. There are actually two error conditions here, and the current implementation only logs one. I've updated the code to log both.